src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala

```diff
@@ -51,38 +57,81 @@ class HttpGoogleServicesDAO(
-        insertGroup(groups,newGroup(bucketName,workspaceName,WorkspaceAccessLevel.Owner)) {
-          val ownersGroupId = toGroupId(bucketName,WorkspaceAccessLevel.Owner)
+
+    // tries to delete all groups expected to have been created whether they actually were or not
+    def rollbackGroups(t: Throwable): Throwable = {
+      Future.traverse(groupAccessLevelsAscending) { accessLevel =>
+        Future {
+          val inserter = groups.delete(newGroup(bucketName, workspaceName, accessLevel).getEmail)
+          blocking {
+            inserter.execute
+          }
+        }
+      }
+      t
+    }
+
+    def insertGroups(workspaceName: WorkspaceName, bucketName: String): Future[Seq[Try[Group]]] = {
+      Future.traverse(groupAccessLevelsAscending) { accessLevel =>
+        toFutureTry(retry(when500) {
+          () => Future {
+            val inserter = groups.insert(newGroup(bucketName, workspaceName, accessLevel))
+            blocking {
+              inserter.execute
+            }
+          }
+        })
+      }
+    }
+
+    def assertSuccessfulGroupInserts: (Seq[Try[Group]]) => Future[Seq[Group]] = { tryGroups =>
+      Future {
+        tryGroups map {
+          _ match {
+            case Success(g) => g
+            case Failure(t) => throw t
+          }
+        }
+      }
+    }
+
+    def insertOwnerMember: (Seq[Group]) => Future[Member] = { _ =>
+      retry(when500) {
+        () => Future {
+          val ownersGroupId = toGroupId(bucketName, WorkspaceAccessLevel.Owner)
           val owner = new Member().setEmail(userInfo.userEmail).setRole(groupMemberRole)
-          val ownerInserter = directory.members.insert(ownersGroupId,owner)
-          retry(ownerInserter.execute,when500)
-
-          val readersEntity = makeGroupEntityString(toGroupId(bucketName,WorkspaceAccessLevel.Read))
-          val writersEntity = makeGroupEntityString(toGroupId(bucketName,WorkspaceAccessLevel.Write))
-          val ownersEntity = makeGroupEntityString(ownersGroupId)
-          val bucket = new Bucket().setName(bucketName).setAcl(List(
-              newBucketAccessControl(readersEntity,WorkspaceAccessLevel.Read),
-              newBucketAccessControl(writersEntity,WorkspaceAccessLevel.Write),
-              newBucketAccessControl(ownersEntity,WorkspaceAccessLevel.Owner)
-            )).setDefaultObjectAcl(List(
-              newObjectAccessControl(readersEntity,WorkspaceAccessLevel.Read),
-              // NB: writers have read access to objects -- there is no write access, objects are immutable
-              newObjectAccessControl(writersEntity,WorkspaceAccessLevel.Read/*sic!*/),
-              newObjectAccessControl(ownersEntity,WorkspaceAccessLevel.Owner)
-            ))
-          val inserter = getStorage(getBucketCredential(userInfo)).buckets.insert(projectId,bucket)
-          retry(inserter.execute,when500)
+          val ownerInserter = directory.members.insert(ownersGroupId, owner)
```

Comment:
I prefer variables to be scoped as tightly as possible. I feel like moving it up interrupts the continuity. "Recomputing" is probably milliseconds only when there is a retry for an operation that is probably orders of magnitude slower.